subprojects/messaging/src/test/groovy/org/gradle/internal/serialize/DefaultSerializerRegistrySpec.groovy

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.serialize
+
+import spock.lang.Specification
+import spock.lang.Subject
+import spock.lang.Timeout
+import spock.util.concurrent.AsyncConditions
+
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+
+class DefaultSerializerRegistrySpec extends Specification {
+    def numThreads = 10
+    List<Class<?>> serializableClasses = [
+        Serializable1, Serializable2, Serializable3, Serializable4, Serializable5, Serializable6, Serializable7, Serializable8, Serializable9, Serializable10,
+        Serializable11, Serializable12, Serializable13, Serializable14, Serializable15, Serializable16, Serializable17, Serializable18, Serializable19, Serializable20,
+        Serializable21, Serializable22, Serializable23, Serializable24, Serializable25, Serializable26, Serializable27, Serializable28, Serializable29, Serializable30,
+        Serializable31, Serializable32, Serializable33, Serializable34, Serializable35, Serializable36, Serializable37, Serializable38, Serializable39, Serializable40,
+        Serializable41, Serializable42, Serializable43, Serializable44, Serializable45, Serializable46, Serializable47, Serializable48, Serializable49, Serializable50,
+        Serializable51, Serializable52, Serializable53, Serializable54, Serializable55, Serializable56, Serializable57, Serializable58, Serializable59, Serializable60,
+        Serializable61, Serializable62, Serializable63, Serializable64, Serializable65, Serializable66, Serializable67, Serializable68, Serializable69, Serializable70,
+        Serializable71, Serializable72, Serializable73, Serializable74, Serializable75, Serializable76, Serializable77, Serializable78, Serializable79, Serializable80,
+        Serializable81, Serializable82, Serializable83, Serializable84, Serializable85, Serializable86, Serializable87, Serializable88, Serializable89, Serializable90,
+        Serializable91, Serializable92, Serializable93, Serializable94, Serializable95, Serializable96, Serializable97, Serializable98, Serializable99, Serializable100
+    ]
+
+    def latch = new CountDownLatch(1)
+
+    @Subject
+    def defaultSerializerRegistry = new DefaultSerializerRegistry()
+
+    @Timeout(15)
+    def "invoking DefaultSerializerRegistry from multiple threads"() {
```

Comment:
I think there should be an integration test which shows the problem. We have many classes which are not thread safe and they shouldn't have to be.

Reply:
I have no idea how to reproduce the issue. I recognize this may be treating a symptom instead of the larger issue, but I don't have enough context nor time to try to solve this issue at that scope.

Reply:
It seems like having multiple tasks which declare custom, serializable types as `@Input` should probably reproduce the problem when you run all those tasks in parallel. They'd need to be part of different projects and the build would probably need to run with `--parallel` so it would work.